cpp/src/parquet/statistics.cc

```diff
@@ -494,6 +494,8 @@ class TypedStatisticsImpl : public TypedStatistics<DType> {
                       int64_t null_count, int64_t distinct_count, bool has_min_max,
                       bool has_null_count, bool has_distinct_count, MemoryPool* pool)
       : TypedStatisticsImpl(descr, pool) {
+    has_null_count_ = has_null_count;
+    has_distinct_count_ = has_distinct_count;
```

Comment:
This is related to the GLib failure.
If this is a correct change, we can change the GLib test:

```diff
diff --git a/c_glib/test/parquet/test-statistics.rb b/c_glib/test/parquet/test-statistics.rb
index 4f21ebf00c..0367084c88 100644
--- a/c_glib/test/parquet/test-statistics.rb
+++ b/c_glib/test/parquet/test-statistics.rb
@@ -51,7 +51,7 @@ class TestParquetStatistics < Test::Unit::TestCase
 
   test("#has_n_distinct_values?") do
     assert do
-      @statistics.has_n_distinct_values?
+      not @statistics.has_n_distinct_values?
     end
   end
 
```

Reply:
It seems that this is a correct change.
How about adding tests related to this to `cpp/src/parquet/statistics_test.cc` or spliting this to another pull request?

Reply:
#34351/34355 is related.

Reply:
I meet the same problem here, I think the syntax of "has_xxx" is like that, for a writer:
* Writer can assure that if has right null-count ( if it not has any bugs )
* Currently I found that ndv is never collected. If a user collect ndv in page1, but not collect ndv in page 2, it should be abandon.

For reader:
* When deserialize, reader should assume that ndv and null_count can be unset ( but currently, it doesn't work like this)
* Deserialized statistics can call merge, but if either `null_count` or `ndv` is unset, all null_count should be discarded.

Reply:
@kou Thanks for the information! Let's solve it in a separate PR because removing the change in statistics does not affect this one.

@mapleFU As you are fixing statistics and hitting the same issue, do you think solving them all together is a good idea? I have left some comments in your PR and let's discuss there. I'll rebase my PR once yours is merged. Thanks!